### PR TITLE
Add copy feature

### PR DESCRIPTION
**Pull request by mrstark14**

### CODE REVIEW

File: views/file.py [hunk 1]
```diff
@@ -280,6 +280,54 @@ def update_shared_users(self, request, *args, **kwargs):
...
+        except File.DoesNotExist:
+            return HttpResponse('file not found', status=status.HTTP_400_BAD_REQUEST)
+        try:
+            folder = Folder.objects.get(pk=destination)
+        except Folder.DoesNotExist:
+            return HttpResponse('destination folder not found', status=status.HTTP_400_BAD_REQUEST)
+        self.check_object_permissions(self.request, folder)
+        try:
+            file_manager = folder.filemanager
+            person = Person.objects.get(pk=folder.person_id)
```
Comment (gauransh7) [inline views/file.py:300]: There should be a additional check that the person is same as the request.person.
What if the destination was corresponding to a folder which is not owned by request.person.

File: views/folder.py [hunk 2]
```diff
@@ -257,3 +258,57 @@ def get_parent_folders(self, request, pk):
...
+        except Folder.DoesNotExist:
+            return HttpResponse('folder not found', status=status.HTTP_400_BAD_REQUEST)
+        try:
+            final_folder = Folder.objects.get(pk=destination)
+        except Folder.DoesNotExist:
+            return HttpResponse('destination folder not found', status=status.HTTP_400_BAD_REQUEST)
+        self.check_object_permissions(self.request, final_folder)
+        try:
+            file_manager = final_folder.filemanager
+            person = Person.objects.get(pk=final_folder.person_id)
```
Comment (gauransh7) [inline views/folder.py:336]: Same here.

File: views/file.py [hunk 3]
```diff
@@ -280,6 +280,54 @@ def update_shared_users(self, request, *args, **kwargs):
         except Exception as e:
             return HttpResponse(f'Unable to change shared_user due to {e}', status=status.HTTP_400_BAD_REQUEST)
 
+    @action(methods=['post'], detail=False, url_name='copy_file', url_path='copy_file')
+    def copy(self, request):
+        data = request.data
+        file_id = int(data.get('id', None))
+        destination = int(data.get('destination', None))
+        try:
+            file = File.objects.get(pk=file_id)
```
Comment (gauransh7) [inline views/file.py:289]: Add a check for the file ownership. (request.person==file.person)

**Review by gauransh7 — CHANGES_REQUESTED** [review]

File: views/file.py (hunk 1 above)
Comment (mrstark14) [reply views/file.py:300]: at line 296, the final folder has check_object_permissions

**Pull request merged**